### ISSUE TITLE
[Core] util.multiprocessing.pool scheduling inefficiencies, blocking behavior in imap and imap_unordered

### ISSUE THREAD

**Issue by minerharry** [ray-project/ray#29453]
### What happened + What you expected to happen

I was poking around in the ray.util.multiprocessing.pool code and I found that all pool methods schedule chunks to actors purely by cycling the current actor index in the pool. This leads to two issues: first, in all iterable-input methods (map+, starmap+, imap+), since the items (or chunks) are assigned deterministically, bad random patterns in the input can lead to one actor in the pool lagging significantly behind:

```python
import time
import ray
from ray.util.multiprocessing import Pool 

ray.init();
num_procs = 4;
p = Pool(num_procs);

lengths = [1,2,1,10,3,1,4,10,1,3,2,10]; ## put long tasks all on one actor due to cyclic nature; actor 4 will always be running the 10s tasks

print("ideal time:",sum(lengths)/num_procs);

start = time.time();


def wait_time(t):
    print(f"waiting {t} seconds...",f"@{time.time()-start}")
    time.sleep(t);
    print(f"done!",f"@{time.time()-start}");
res = p.imap(wait_time,lengths,chunksize=1);
list(res); #clear queue for iterators

end = time.time();

print("actual time:",end-start);
```

outputs
```
ideal time: 12.0
(PoolActor pid=24560) waiting 1 seconds... @0.014564990997314453
(PoolActor pid=31596) waiting 10 seconds... @0.016566753387451172
(PoolActor pid=4932) waiting 1 seconds... @0.012836456298828125
(PoolActor pid=2624) waiting 2 seconds... @0.010305643081665039
(PoolActor pid=24560) done! @1.0175859928131104
(PoolActor pid=4932) done! @1.0175859928131104
(PoolActor pid=4932) waiting 3 seconds... @1.0259771347045898
(PoolActor pid=24560) waiting 4 seconds... @2.0303306579589844
(PoolActor pid=2624) done! @2.025196075439453
(PoolActor pid=2624) waiting 1 seconds... @2.0277397632598877
(PoolActor pid=2624) done! @3.0369625091552734
(PoolActor pid=4932) done! @4.031100273132324
(PoolActor pid=4932) waiting 1 seconds... @4.035675048828125
(PoolActor pid=4932) done! @5.042609691619873
(PoolActor pid=24560) done! @6.034060716629028
(PoolActor pid=24560) waiting 2 seconds... @6.0408055782318115
(PoolActor pid=2624) waiting 3 seconds... @6.039994955062866
(PoolActor pid=24560) done! @8.042097806930542
(PoolActor pid=2624) done! @9.05439305305481
(PoolActor pid=31596) done! @10.017773389816284
(PoolActor pid=31596) waiting 10 seconds... @10.019720077514648
(PoolActor pid=31596) done! @20.023097038269043
(PoolActor pid=31596) waiting 10 seconds... @20.025118827819824
actual time: 30.039400577545166
(PoolActor pid=31596) done! @30.03778886795044
```
with similar results using .map(), .imap_unordered(), and .starmap(). 
Even though three actors (2624, 2456, and 4392) are done with all tasks by 10sec, since the 10s tasks were all multiples of four, they get assigned to actor of pid 31596, which executes tasks for a total of 30s. The ideal time is of course likely unrealistic even with optimal scheduling given the order and distribution of tasks, but the current behavior is still quite bad.

Second, in imap and imap_unordered, a similar effect can be observed (though sort of backwards) because the chunks can't all be submitted at once, meaning future actors in the cycle are waiting on earlier actors to complete their chunks.

```python
import time
import ray
from ray.util.multiprocessing import Pool 

ray.init();
num_procs = 4;
p = Pool(num_procs);

lengths = [30,1,1,1,1,1,1,1,1,1,1,1]; ## put long task on first actor; blocks other actors from receiving subsequent tasks

print("ideal time:",sum(lengths)/num_procs);

start = time.time();


def wait_time(t):
    print(f"waiting {t} seconds...",f"@{time.time()-start}")
    time.sleep(t);
    print(f"done!",f"@{time.time()-start}");
    time.sleep(0.5);
res = p.imap_unordered(wait_time,lengths,chunksize=1);
list(res); #clear queue for iterators

end = time.time();

print("actual time:",end-start);
```
(Note that in this version there's an extra time.sleep() after the "done" print to account for the slight delay in writing to stdout; output looked a little messy without it.)

And the output:
```
ideal time: 10.25
(PoolActor pid=13320) waiting 1 seconds... @0.009492158889770508
(PoolActor pid=33356) waiting 30 seconds... @0.005202531814575195
(PoolActor pid=34340) waiting 1 seconds... @0.01501607894897461
(PoolActor pid=33612) waiting 1 seconds... @0.013006925582885742
(PoolActor pid=13320) done! @1.0130796432495117
(PoolActor pid=34340) done! @1.029155969619751
(PoolActor pid=33612) done! @1.029155969619751 
# NOTE TIME SKIP
(PoolActor pid=33356) done! @30.017018795013428
(PoolActor pid=13320) waiting 1 seconds... @30.53541088104248
(PoolActor pid=33356) waiting 1 seconds... @30.534412384033203
(PoolActor pid=34340) waiting 1 seconds... @30.53541088104248
(PoolActor pid=33612) waiting 1 seconds... @30.53641414642334
(PoolActor pid=13320) done! @31.5425968170166
(PoolActor pid=33356) done! @31.542810916900635
(PoolActor pid=34340) done! @31.542810916900635
(PoolActor pid=33612) done! @31.5425968170166
(PoolActor pid=13320) waiting 1 seconds... @32.06012225151062
(PoolActor pid=33356) waiting 1 seconds... @32.059123039245605
(PoolActor pid=34340) waiting 1 seconds... @32.06119704246521
(PoolActor pid=33612) waiting 1 seconds... @32.06119704246521
(PoolActor pid=13320) done! @33.06329393386841
(PoolActor pid=33356) done! @33.06329393386841
(PoolActor pid=34340) done! @33.06329393386841
(PoolActor pid=33612) done! @33.06329393386841
actual time: 33.57944679260254
```
"# NOTE TIME SKIP" added after the fact to point out where all processes were waiting on actor pid 33356 to finish the 30 second task. Although the ideal time would never be achievable with even the best scheduling because of the mismatch of task lengths, if lengths were something like [29,1,1,1,1,29,29,29], the program would take 60s instead of the more ideal 30s. This behavior is also seen with lengths of [1,1,1,1,30,1,1,1,1,1,1] and [1,1,30,1,1,1,1,1,1,1,1]

Both of these issues could be fixed by assigning tasks to only free actors. NOTE: I'm not exactly sure why this second behavior is blocking in imap and imap_unordered; from testing, it seems that the ray.wait in the ResultThread.run method is only returning when the long process is finished. I'll open a separate issue for that after I run more tests. However, when I replaced the `actor_index = len(self._submitted_chunks) % len(self._pool._actor_pool)` line in _submit_next_chunk with `actor_index = random.randrange(len(self._pool._actor_pool))`, the issue was bypassed for the majority of task elements. If a new task was assigned to the long actor, it would sometimes still delay the other tasks, but with nonrandom assignment the issue would be fixed.

Ideally, this task assignment would be with something akin to the scheduling protocols used for tasks and ray.remote(); however, I think it could be as simple as a ray.wait call:

```python
    def _get_next_free_actor_index(self):
        ref_map = {actor[0].ping.remote():(actor[1],k) for k,actor in enumerate(self._pool._actor_pool)}; #objref:count,actorindex
        ready, _ = ray.wait(list(ref_map.keys()));
        ready_actors = [ref_map[ref] for ref in ready];
        ready_index = min(ready_actors,key=lambda r:r[0])[1]; #get ready actor with fewest tasks
        return ready_index;
```
Alternatively, if that runs into similar issues to imap and imap_unordered, using a ray.util.queue would also be possible (the way python's multiprocessing.pool does it) with each worker checking for new tasks once it's executed its own, though that would mean more code restructuring
    

    

### Versions / Dependencies

Ray version: 3.0.0.dev0 (nightly build accessed on 10/17)
Python Version: 3.10
OS: Windows 10

### Reproduction script

inefficient scheduling:
```python
import time
import ray
from ray.util.multiprocessing import Pool 

ray.init();
num_procs = 4;
p = Pool(num_procs);

lengths = [1,2,1,10,3,1,4,10,1,3,2,10]; ## put long tasks all on one actor due to cyclic nature; actor 4 will always be running the 10s tasks

print("ideal time:",sum(lengths)/num_procs);

start = time.time();


def wait_time(t):
    print(f"waiting {t} seconds...",f"@{time.time()-start}")
    time.sleep(t);
    print(f"done!",f"@{time.time()-start}");
res = p.imap(wait_time,lengths,chunksize=1);
list(res); #clear queue for iterators

end = time.time();

print("actual time:",end-start);
```
imap and imap_unordered blocking tasks:
```python
import time
import ray
from ray.util.multiprocessing import Pool 

ray.init();
num_procs = 4;
p = Pool(num_procs);

lengths = [30,1,1,1,1,1,1,1,1,1,1,1]; ## put long task on first actor; blocks other actors from receiving subsequent tasks

print("ideal time:",sum(lengths)/num_procs);

start = time.time();


def wait_time(t):
    print(f"waiting {t} seconds...",f"@{time.time()-start}")
    time.sleep(t);
    print(f"done!",f"@{time.time()-start}");
    time.sleep(0.5);
res = p.imap_unordered(wait_time,lengths,chunksize=1);
list(res); #clear queue for iterators

end = time.time();

print("actual time:",end-start);
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by cadedaniel** [comment]
Thanks for the details and writeup! Let me see if we have a current plan for this.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
